keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+When a user start up a new etcd cluster, it is unnecessary and confusing to require them to know the difference between server level and cluster level features. It is really up to the developers to register them to the proper type in the code.
+
+So the cluster feature gates for a server can be set in the same way as the server level feature gates: with the `--feature-gates`flag or in the `config-file` during startup. We do not support dynamically changing the feature gates when the server is running.
+
+After starting up, the values of all features (server+cluster level) will be stored in the `ServerFeatureGate` variable of the server.
+
+### Consensus Algorithm
+
+To guarantee consistent value of if a feature is enabled in the whole cluster, the leader would decide if the feature is enabled for all cluster members, and propagate the information to all members. 
+
+1. When an etcd server starts, the values of all cluster level features in its `ServerFeatureGate` would saved in the [`member.Attributes`](https://github.com/etcd-io/etcd/blob/e37a67e40b3f5ff8ef81f9de6e7f475f17fda32b/server/etcdserver/api/membership/member.go#L38) as the field `Attributes.feature_gates`. (see the [discussion](#push-vs-poll-when-leader-decides-cluster-feature-gate) about why we choose to push the information through raft)
+
+1. The attributes would then be [published through raft](https://github.com/etcd-io/etcd/blob/e37a67e40b3f5ff8ef81f9de6e7f475f17fda32b/server/etcdserver/server.go#L1745) and stored in the `members` bucket in the backend of all members. Whenever a new member joins or an existing member restarts, its feature gate attributes would be automatically updated in the start up process.
+
+1. After a leader is elected, the leader will decide the values for the cluster level features based on the cluster version and server feature values of all members it receives.
+
+1. The leader sends the final feature values in `ClusterFeatureGateSetRequest` through raft. 
+
+1. Each member apply the updates to their cluster feature gate, and save the results in the `cluster` bucket in the backend.
+
+#### New Raft Proto Changes
+
+```proto
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+
+message Attributes {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  string name = 1;
+  repeated string client_urls = 2;
+  // the values of all cluster level feature gates set by the configuration of the member server.
+  Feature proposed_feature_gates = 3;
+}
+
+message ClusterFeatureGateSetRequest {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  // the values of all cluster level feature gates for the whole cluster.
+  Feature feature_gates = 1;
+}
+```
+
+#### How to Reach Consensus of Feature Gates for a New Cluster
+
+![setting cluster feature gate for a new cluster](./cluster_feature_gate_new.png "setting cluster feature gate for a new cluster")
+
+We will leverage the consensus `ClusterVersion` to negotiate the cluster level feature gate values:
```

Comment:
I think we should consider reconciling cluster version and it's features in one proto to ensure they are consistent

Reply:
@siyuanfoundation what do you think about having cluster version and feature flags to be atomic operation?

Reply:
There are some problems with reconciling cluster version and it's features in one proto. Different members can have different defaults for a feature gate at a certain version, and they do not know what values to propose until the cluster version is known. So I have separated the algorithm to 2 steps: 
1. cluster version is set along with the feature gate default values at that cluster version in 1 request.
2. members propose their own feature gate values at the the established cluster version, and leader reconcile it.